src/data/nat/factorization/basic.lean

```diff
@@ -481,6 +481,13 @@ begin
   { simp [hqp, (factorization_le_iff_dvd hd0 hn0).2 hdn q] },
 end
 
+/-- If `n` is a nonzero natural number and `p` is a prime, then there are natural numbers `e`
+and `n'` such that `n'` is not divisible by `p` and `n = p^e * n'`. -/
+lemma exists_prime_pow_and_compl {n : ℕ} (hn : n ≠ 0) (p : ℕ) [fact p.prime] :
```

Comment:
This does not require `p` to be prime, right? It seems essentially [exists_eq_pow_mul_and_not_dvd](https://leanprover-community.github.io/mathlib_docs/ring_theory/multiplicity.html#multiplicity.exists_eq_pow_mul_and_not_dvd).

Reply:
That `p` is prime is needed for the proof (via `[not_dvd_ord_compl](https://leanprover-community.github.io/mathlib_docs/data/nat/factorization/basic.html#nat.not_dvd_ord_compl)`). The result you link to looks similar, but I'm a bit concerned regarding the `[multiplicity.finite](https://leanprover-community.github.io/mathlib_docs/ring_theory/multiplicity.html#multiplicity.finite) a b` requirement there. Is there a way to obtain that easily here? Anyway, for applications with natural numbers, assuming `p` prime is very natural in this context.

Reply:
Obtaining `multiplicty.finite` should follow if you only assume `p ≠ 1`. See [multiplicity.finite_nat_iff](https://leanprover-community.github.io/mathlib_docs/ring_theory/multiplicity.html#multiplicity.finite_nat_iff).

Reply:
Sure, you need `p ≠ 1`.

Reply:
So what do you suggest?
* Use [exists_eq_pow_mul_and_not_dvd](https://leanprover-community.github.io/mathlib_docs/ring_theory/multiplicity.html#multiplicity.exists_eq_pow_mul_and_not_dvd) for the proof here (perhaps with a weakened hypothesis on `p`, which would, however, necessitate inserting a proof that `p ≠ 1` where I use it)
* Use [exists_eq_pow_mul_and_not_dvd](https://leanprover-community.github.io/mathlib_docs/ring_theory/multiplicity.html#multiplicity.exists_eq_pow_mul_and_not_dvd) directly in my Jacobi symbol code (or, alternatively, inline the existing proof) and remove the lemma
* None of the above

Reply:
I think this lemma should assume `p ≠ 1`.
If you have `hp : fact p.prime` then `hp.out.ne_one : p ≠ 1` doesn't seem too bad.

Reply:
I don't have a real preference here, it is a reasonable lemma to have for `nat`, without importing all the multiplicity stuff, so you can also prove it directly. But in any case we try to avoid unnecessary assumption (like `prime p` here).

Reply:
When I try to change the proof using the multiplicity stuff:
```
lemma exists_prime_pow_and_compl {n : ℕ} (hn : n ≠ 0) (p : ℕ) (hp : p ≠ 1) :
  ∃ e n' : ℕ, ¬ p ∣ n' ∧ n = p ^ e * n' :=
begin
  have hfin := multiplicity.finite_nat_iff.mpr ⟨hp, nat.pos_of_ne_zero hn⟩,
  obtain ⟨a', h₁, h₂⟩ := multiplicity.exists_eq_pow_mul_and_not_dvd hfin,
  refine ⟨(multiplicity n p).get _, a', h₂, h₁⟩,
end
```
I get the error
```
term
  h₁
has type
  n = p ^ (multiplicity p n).get hfin * a'
but is expected to have type
  n = p ^ (multiplicity n p).get ?m_1 * a'
```
Apparently, `.get` wants something of type `.dom`, and I have no idea how to convert `hfin` suitably. I'm confused, since in the statement of `multiplicity.exists_eq_pow_mul_and_not_dvd`, something of type `hfin` is used in exactly the same position.

Reply:
Hmm, `refine ⟨_, a', h₂, h₁⟩` does it. Still, I don't quite understand what is going on here...

Reply:
I've now pushed the new proof.

Reply:
> 

You swaped `n` and `p`.
```
lemma exists_prime_pow_and_compl {n : ℕ} (hn : n ≠ 0) (p : ℕ) (hp : p ≠ 1) :
  ∃ e n' : ℕ, ¬ p ∣ n' ∧ n = p ^ e * n' :=
begin
  have hfin := multiplicity.finite_nat_iff.mpr ⟨hp, nat.pos_of_ne_zero hn⟩,
  obtain ⟨a', h₁, h₂⟩ := multiplicity.exists_eq_pow_mul_and_not_dvd hfin,
  exact ⟨(multiplicity p n).get _, a', h₂, h₁⟩,
end
```
works.